tests/integration/preserve_ip_config_test.py

```diff
@@ -0,0 +1,136 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+import gi                           # pylint: disable=import-error
+gi.require_version('NM', '1.0')     # NOQA: F402
+from gi.repository import NM        # pylint: disable=import-error
+from libnmstate import netapplier
+
+from .testlib import statelib
+from .testlib import cmd as libcmd
+from .testlib.statelib import INTERFACES
+
+
+_EXTRA_CONFIGS = [
+    {
+        'name': 'ipv4.dad-timeout',
+        'value': '1000'
+    },
+    {
+        'name': 'ipv6.dhcp-hostname',
+        'value': 'libnmstate.example.com'
+    }
+]
+
+# TEST-NET addresses: https://tools.ietf.org/html/rfc5737#section-3
+IPV4_ADDRESS1 = '192.0.2.251'
+# IPv6 Address Prefix Reserved for Documentation:
+# https://tools.ietf.org/html/rfc3849
+IPV6_ADDRESS1 = '2001:db8:1::1'
+
+
+def _get_nm_profile_uuid(iface_name):
+    nmcli = NM.Client.new(None)
+    cur_dev = None
+    for dev in nmcli.get_all_devices():
+        if dev.get_iface() == iface_name:
+            cur_dev = dev
+            break
+
+    assert cur_dev
+
+    active_conn = cur_dev.get_active_connection()
+    assert active_conn
+    return active_conn.get_uuid()
+
+
+def test_reapply_preserve_ip_config():
+    netapplier.apply(
+        {
+            'interfaces': [
+                {
+                    'name': 'eth1',
+                    'type': 'ethernet',
+                    'state': 'up',
+                    'ipv4': {
+                        'address': [
+                            {
+                                'ip': IPV4_ADDRESS1,
+                                'prefix-length': 24
+                            }
+                        ],
+                        'enabled': True
+                    },
+                    'ipv6': {
+                        'address': [
+                            {
+                                'ip': IPV6_ADDRESS1,
+                                'prefix-length': 64
+                            }
+                        ],
+                        'enabled': True
+                    },
+                    'mtu': 1500
+                },
+            ]
+        })
+    cur_state = statelib.show_only(('eth1',))
+    iface_name = cur_state[INTERFACES][0]['name']
+
+    uuid = _get_nm_profile_uuid(iface_name)
+    old_cfg = _get_cur_extra_ip_config(uuid)
+
+    try:
+        _apply_extra_ip_config(uuid, _EXTRA_CONFIGS)
+        netapplier.apply(cur_state)
+        _assert_extra_ip_config(uuid, _EXTRA_CONFIGS)
+    except Exception:
+        _apply_extra_ip_config(uuid, old_cfg)
+        raise
+
+    _apply_extra_ip_config(uuid, old_cfg)
+
+
+def _get_cur_extra_ip_config(uuid):
+    props = [c['name'] for c in _EXTRA_CONFIGS]
+    (rc, output, _) = libcmd.exec_cmd(['nmcli', '--get-values',
+                                       ','.join(props), 'connection', 'show',
+                                       uuid])
+    assert rc == 0
+    return [
```

Comment:
Why do we need this structure? It would be less messy with a direct approach of fetching and setting individual properties.

Reply:
I use that to avoid using magic string for the property name `ipv4.dad-timeout` and `ipv6.dhcp-hostname`. Do you have better approach does not require hard code magic string?

Reply:
I think that packaging the two in a dict structure is not serving anything. Each should be self contained not coupled together.